Cesium3DTilesSelection/src/TerrainQuery.cpp

```diff
@@ -0,0 +1,125 @@
+
+#include "TileUtilities.h"
+#include "TilesetContentManager.h"
+
+#include <CesiumGeometry/IntersectionTests.h>
+#include <CesiumGeospatial/GlobeRectangle.h>
+#include <CesiumGltfContent/GltfUtilities.h>
+
+using namespace Cesium3DTilesSelection;
+using namespace CesiumGeospatial;
+using namespace CesiumGeometry;
+using namespace CesiumUtility;
+using namespace CesiumAsync;
+
+namespace {
+bool boundingVolumeContainsCoordinate(
+    const BoundingVolume& boundingVolume,
+    const Ray& ray,
+    const Cartographic& coordinate) {
+  struct Operation {
+    const Ray& ray;
+    const Cartographic& coordinate;
+
+    bool operator()(const OrientedBoundingBox& boundingBox) noexcept {
+      std::optional<double> t =
+          IntersectionTests::rayOBBParametric(ray, boundingBox);
+      return t && t.value() >= 0;
+    }
+
+    bool operator()(const BoundingRegion& boundingRegion) noexcept {
+      return boundingRegion.getRectangle().contains(coordinate);
+    }
+
+    bool operator()(const BoundingSphere& boundingSphere) noexcept {
+      std::optional<double> t =
+          IntersectionTests::raySphereParametric(ray, boundingSphere);
+      return t && t.value() >= 0;
+    }
+
+    bool operator()(
+        const BoundingRegionWithLooseFittingHeights& boundingRegion) noexcept {
+      return boundingRegion.getBoundingRegion().getRectangle().contains(
+          coordinate);
+    }
+
+    bool operator()(const S2CellBoundingVolume& s2Cell) noexcept {
+      return s2Cell.computeBoundingRegion().getRectangle().contains(coordinate);
+    }
+  };
+
+  return std::visit(Operation{ray, coordinate}, boundingVolume);
+}
+
+} // namespace
+
+void TerrainQuery::intersectVisibleTile(Tile* pTile) {
+  TileRenderContent* pRenderContent = pTile->getContent().getRenderContent();
+  if (!pRenderContent)
+    return;
+
+  auto gltfIntersectResult =
+      CesiumGltfContent::GltfUtilities::intersectRayGltfModel(
+          this->ray,
+          pRenderContent->getModel(),
+          true,
+          pTile->getTransform());
+
+  if (!gltfIntersectResult.hit.has_value())
+    return;
+
+  // Set ray info to this hit if closer, or the first hit
+  if (!this->intersectResult.hit.has_value()) {
+    this->intersectResult = std::move(gltfIntersectResult);
+  } else {
+    double prevDistSq = this->intersectResult.hit->rayToWorldPointDistanceSq;
+    double thisDistSq = intersectResult.hit->rayToWorldPointDistanceSq;
+    if (thisDistSq < prevDistSq)
+      this->intersectResult = std::move(gltfIntersectResult);
+  }
+}
+
+void TerrainQuery::findCandidateTiles(
+    Tile* pTile,
+    std::vector<std::string>& warnings) {
+
+  // If tile failed to load, this means we can't complete the intersection
+  if (pTile->getState() == TileLoadState::Failed) {
+    warnings.push_back("Tile load failed during query. Ignoring.");
+    return;
+  }
+
+  if (pTile->getChildren().empty()) {
+    // This is a leaf node, add it to the list
+    candidateTiles.push_back(pTile);
```

Comment:
What if it's only temporarily a leaf node? For example, it's an external tileset that hasn't been loaded yet. Some mechanism will need to ensure a) the external tileset gets loaded, and b) we don't consider the height available until we load it and then potentially load some of the tiles it contains. Is there already such a mechanism in place? Will it also work with implicit tilesets that need to load subtrees? With quantized-mesh where tiles are created on demand based on availability, and availability is embedded in tiles at certain levels?

Reply:
This case should be handled.

The general flow is `TerrainQuery::findCandidateTiles` first to gather the potential tiles we need to intersect. If any of these aren't loaded, we wait for them to get loaded, then call `TerrainQuery::findCandidateTiles` again to get the final set of loaded nodes (I assume temporary nodes are no longer in here). Then all these nodes pass through `TerrainQuery::intersectVisibleTile`.

I suspect these two functions might continue to be modified in the future. There may be more edge cases where the traversal isn't right and we may want to support more progressive intersect traversals too, rather than just waiting until the most detailed tiles are completely loaded.

Reply:
How to use in unity?

Reply:
Hi @killjin,

This PR is for adding the basic functionality in Cesium Native. Our next steps would include designing the API + user experience, to expose this to our runtimes.

I don't have a time estimate for when this will be completed, but you're always welcome to modify the plugin yourself to access these features early. :smile: